// htmlbars/dist/cjs/htmlbars-runtime/expression-visitor.js
exports.__esModule = true;
exports.acceptParams = acceptParams;
exports.acceptHash = acceptHash;
/**
  # Expression Nodes:

  These nodes are not directly responsible for any part of the DOM, but are
  eventually passed to a Statement Node.

  * get
  * subexpr
  * concat
*/

function acceptParams(nodes, env, scope) {
  var array = [];

  for (var i = 0, l = nodes.length; i < l; i++) {
    array.push(acceptExpression(nodes[i], env, scope).value);
  }

  return array;
}

function acceptHash(pairs, env, scope) {
  var object = {};

  for (var i = 0, l = pairs.length; i < l; i += 2) {
    var key = pairs[i];
    var value = pairs[i + 1];
    object[key] = acceptExpression(value, env, scope).value;
  }

  return object;
}

function acceptExpression(node, env, scope) {
  var ret = { value: null };

  // Primitive literals are unambiguously non-array representations of
  // themselves.
  if (Array.isArray(node)) {
    // if (node.length !== 7) { throw new Error('FIXME: Invalid statement length!'); }

    ret.value = evaluateNode(node, env, scope);
  } else {
    ret.value = node;
  }

  return ret;
}

function evaluateNode(node, env, scope) {
  switch (node[0]) {
    // can be used by manualElement
    case 'value':
      return node[1];
    case 'get':
      return evaluateGet(node, env, scope);
    case 'subexpr':
      return evaluateSubexpr(node, env, scope);
    case 'concat':
      return evaluateConcat(node, env, scope);
  }
}

function evaluateGet(node, env, scope) {
  var path = node[1];

  return env.hooks.get(env, scope, path);
}

function evaluateSubexpr(node, env, scope) {
  var path = node[1];
  var rawParams = node[2];
  var rawHash = node[3];

  var params = acceptParams(rawParams, env, scope);
  var hash = acceptHash(rawHash, env, scope);

  return env.hooks.subexpr(env, scope, path, params, hash);
}

function evaluateConcat(node, env, scope) {
  var rawParts = node[1];

  var parts = acceptParams(rawParts, env, scope);

  return env.hooks.concat(env, parts);
}
